### PR TITLE
InnerHit new implementation

### PR DESCRIPTION
**Pull request by tchiotludo**
Hi, 

It's not a full pull request but more a implementation question that need some more work to be completed (tests).
InnerHits support mostly all parameters available for search (from, size, highlight, source, ...)
The current implementation as partial support for this parameters (from some via addParameters).
My proposal is to give a full search to innerhits, It's more user friendly.

In case you agree, you can close this one : https://github.com/ongr-io/ElasticsearchDSL/pull/153 that will be duplicate.

What do you think about it ?

### CODE REVIEW

**Comment by saimaz** [comment]
Well, `Search` object scope is very wide. Are you sure that `inner_hits` needs child `inner_hist` or `suggest` or something other what's is in `Search`. That's why we added a type hint with builder interface, where you can add anything from query scope.

**Comment by tchiotludo** [comment]
BuilderInterface does not help in our case, here is a example of valid query : 
```json
{
    "inner_hits": {
        "article": {
            "type": {
                "article": {
                    "sort": {
                        "article_id": {}
                    },
                    "size": 10
                }
            }
        }
    }
}
``` 
since you forced a query : 
```php
'query' => $this->getQuery() ? $this->getQuery()->toArray() : null,
```
It will send "query" instead of "sort"

to deal with, we must use this tricky syntax : 
```php
$query->addInnerHit($parent = (new ParentInnerHit('article', 'article')));
$parent->addParameter('size', 10);
$parent->addParameter('sort', (new FieldSort('article_id'))->toArray());
```
This works but seems tricky 

Using search will allow ide autocompletion but you are true, not all search functionality is allow in innerHits.
- Simple solution, in setQuery, add some control and throw exception.
- Complexe one : make `class Search` use a lot of trait, and create a InnerHitsSearch class that use this trait

**Comment by saimaz** [comment]
I see your point. Well in this particular case change `BuilderInterface` to `Search`makes sense. Then it's up to the user what he adds to the query. Moreover, elasticsearch will throw an exception if something is wrong.

**Comment by tchiotludo** [comment]
Ok so you agree and I can go on the final implementation with test ?

**Comment by saimaz** [comment]
Yes

**Comment by tchiotludo** [comment]
I think it's ok now

File: src/InnerHit/NestedInnerHit.php [hunk 1]
```diff
@@ -84,15 +75,11 @@ public function getQuery()
     }
 
     /**
-     * @param BuilderInterface $query
-     *
-     * @return $this
+     * @param Search $query
      */
-    public function setQuery(BuilderInterface $query)
+    public function setQuery(Search $query = null)
```
Comment (saimaz) [inline src/InnerHit/NestedInnerHit.php]: Please remove default value from setter. It makes no sense to have setter like this.

Should be: 
```php
public function setQuery(Search $query)
{
    $this->query = $query;

    return $this;
}
```

File: src/InnerHit/NestedInnerHit.php [hunk 2]
```diff
@@ -65,14 +60,10 @@ public function getPath()
...
     /**
      * @param string $path
-     *
-     * @return $this
      */
     public function setPath($path)
     {
         $this->path = $path;
-
-        return $this;
```
Comment (saimaz) [inline src/InnerHit/NestedInnerHit.php]: Why you are removing return statement?

**Comment by saimaz** [comment]
Can you append changes in the docs as well according to change with `Search` object?

**Comment by tchiotludo** [comment]
- for the removing of `return $this`, my fault, failed merge.
- for the setter `public function setQuery(Search $query = null)`, same than earlier, query is optional 
- I rename setQuery for setSearch to be consistent

**Comment by saimaz** [comment]
Yes, query is optional in inner hits, but to have a default value in setter doesn't make sense. I already fixed your previous implementation by removing default value. It's very simple, if you don't want to set query do not call setter.

**Comment by saimaz** [comment]
Everything else looks good. Thanks for the support!

**Comment by tchiotludo** [comment]
same debate than earlier, I need to change the construct to  : 

```php 
    /**
     * Inner hits container init.
     *
     * @param string $name
     * @param string $path
     * @param Search $query
     */
    public function __construct($name, $path, Search $query = null)
    {
        $this->setName($name);
        $this->setPath($path);
        if ($query) {
            $this->setSearch($query);
        }
    }
``` 

But earlier, you tell me to put a null value on setter, so I do the same here :smile:

**Comment by saimaz** [comment]
I believe there was a misunderstanding. In case the query is optional maybe it would be better to remove it from the constructor?

**Comment by saimaz** [comment]
btw I have a doubt about `setSearch` and it's parameter `$query` naming. The function name look ok, maybe parameter name should be `$search` as well. In other way around maybe it's not good to say that we are setting here a search object, because main objective is to set a query.

**Comment by tchiotludo** [comment]
- to be consistent, I changed $query properties & parameters to $search
- I remove the default value on settter and add a test on __construct

The search parameter is optional on construct but really needed to have a working innerHits.
In my case, I need to expend the size of inner hits (seems to be 3 by default) and I also need to filter innerHits.
I think only few people will use innerHits with empty search (meaning having top 3 child order by score) so IMO it's better to keep it nullable

Tell me what do you think about that

**Comment by saimaz** [comment]
LGTM

**Pull request merged**